### ISSUE TITLE
`IsPresent` is unsafe to override with respect to `Presence` invalidations

### ISSUE THREAD

**Issue by bdach** [ppy/osu-framework#5129]
Going to make this a full issue as a follow up to ppy/osu#17823 / ppy/osu#17937 for posterity at least, so it doesn't get lost.

Right now, whenever a drawable overrides `IsPresent` to include its own additional criteria for presence (like a filtered state, or `Scheduler.HasPendingTasks`, or similar), for full correctness it should also *manually* ensure that when that additional criteria changes, it should be checked whether the value of `IsPresent` has changed and an appropriate invalidation must be raised so that any other ancestor knows about presence changing. Framework does that manually in many places:

https://github.com/ppy/osu-framework/blob/d84307a4edfd22be382cd8329ecc40fbc33d33c9/osu.Framework/Graphics/Drawable.cs#L1303-L1304

https://github.com/ppy/osu-framework/blob/d84307a4edfd22be382cd8329ecc40fbc33d33c9/osu.Framework/Graphics/Drawable.cs#L979-L980

but out of all of the `IsPresent` overrides that I know of, only one does it (and it is the one fixed in the aforementioned PR). Ideally this should Just Work™, but because `IsPresent` is a computed/derived get-only property, it is a bit difficult to actually be able to tell when it changes without examining if it did on every read - which I guess would be the proposal here, to make sure API is not broken.

**Comment by peppy** [comment]
I do wonder if `IsPresent` should be overridable in the first place. We've had so many issues related to it over the years.

**Comment by bdach** [comment]
I don't recall having so many problems with _overriding_ `IsPresent`, rather than `isPresent` being a frequent source of issues in and of itself. Don't get me wrong, it's a valid and important optimisation, but it has subtle knock-on effects (on schedules, among others) that frequently need working around. Removing the ability to override `IsPresent` would prevent those workarounds from working.

**Comment by smoogipoo** [comment]
Yeah, this is a pretty big pain point of `IsPresent`. Even the current overrides of it adding `Scheduler.HasPendingTasks` are unsafe because the presence check occurs before the scheduler is updated.

I'm not entirely sure how to resolve this in a clean way yet. There's a variety of things that we could try (I was about to post an example here), but I think the best that we may be able to do is to hint that changes to `IsPresent` need to be notified via a presence invalidation, and potentially add hinting to this via a .NET analyser.
